### DISCUSSION TITLE
Does anyone have their graphql service deployed on Azure? Did you need special firewall rules?

### DISCUSSION THREAD

**Discussion by rsdickerson** [urql-graphql/urql#1913] [Q&A] [answered]
We have our GraphQL service deployed on Azure. To get past the WAF we're having to add a rule to ignore 'query'. Without this rule in place we get '403 Forbidden'. With the rule in queries work fine.

While attempting to figure out why WAF is rejecting our graphql queries using Postman to send the query I'm seeing some weird behavior by the WAF.

- rejects if 'id' is first field name of type's fields; for example, user { id name } is rejected while user { name id } works fine
- large queries with CRLF after each field name (which is what urql does) are rejected, but with CRLF replaced with space it works fine

Any ideas? 

Also, is there any way to get urql to not put a CRLF after every field in the query?

**Comment by rsdickerson** [comment, marked as answer]
I found that urql is calling a print command in the GraphQL library prior to sending the request. This print command is formatting the query with newlines and spacing. Azure WAF does not like that. To get past the Azure WAF rules I had to replace all the newlines with spaces. I also removed the extra spacing, but I don't think that is really necessary. I also found that I had to move the 'id' field in a type's selection fields out of the first position. That was just weird, but without doing that Azure WAF still rejected the request.

My solution was to add this code to service-worker.js.

`
const fetchWithTrimmedBody = async ({request}) => {
    let serialized = await serialize(request)
    // Trim body of \n and spaces
    serialized.body = serialized.body.replace(/\\n\s*/g, ' ')
    let deserialized = await deserialize(serialized)
    return await fetch(deserialized)
}

registerRoute(
    ({url, request, event}) => {
        return url.pathname.endsWith('/api/gql')
    },
    fetchWithTrimmedBody,
    'POST'
)
`

Here's the serialize/deserialize functions:
[request.js.txt](https://github.com/FormidableLabs/urql/files/7178096/request.js.txt)

**Reply by JoviDeCroock** [reply]
You can also use https://github.com/kitten/graphql-web-lite if it's on the client by aliasing graphql to graphql-web-lite this doesn't print newlines and saves some bundle size along the way.

**Reply by rsdickerson** [reply]
Thanks. That sounds like a much simpler solution compared to our customization to the Service Worker. I appreciate the info.
